File: recipes/subms-block-cache/rust/examples/growth_main.rs
//! Storage-growth capture for the block cache: stream far more distinct keys
//! through a fixed-capacity LRU than it can hold, and confirm the resident set
//! stays pinned at the capacity. Eviction is what bounds memory here - this
//! proves it, rather than the cache being a map that quietly grows.
//!
//! Emits the stable subms growth JSON on stdout.
//!
//! ```sh
//! cat <<EOF | cargo run --release --example growth_main --features harness
//! rounds=50
//! capacity=1024
//! inserts_per_round=20000
//! EOF
//! ```

use std::collections::BTreeMap;
use std::io::{self, Read};
use std::process::ExitCode;

use subms::{SubMsGrowthClass, SubMsGrowthRecipe, grow, growth_to_json};
use subms_block_cache::BlockCache;

const VALUE_BYTES: usize = 256;
// Rough per-entry heap cost: value + key (u64) + slot/index bookkeeping.
const ENTRY_BYTES: u64 = (VALUE_BYTES + 8 + 32) as u64;

struct CacheChurn {
    cache: BlockCache<u64, Vec<u8>>,
    capacity: usize,
    rounds: usize,
    inserts_per_round: usize,
    value: Vec<u8>,
    next: u64,
}

impl SubMsGrowthRecipe for CacheChurn {
    fn name(&self) -> &str {
        "subms-block-cache"
    }
    fn op_name(&self) -> &str {
        "put"
    }
    fn rounds(&self) -> usize {
        self.rounds
    }
    fn ops_per_round(&self) -> usize {
        self.inserts_per_round
    }
    fn op(&mut self, _round: usize, _i: usize) {
        // A fresh distinct key every op: nothing is ever re-hit, so a naive
        // unbounded map would grow to millions - the LRU must evict to stay flat.
        self.cache.put(self.next, self.value.clone());
        self.next += 1;
    }
    fn memory_bytes(&mut self) -> u64 {
        self.cache.len() as u64 * ENTRY_BYTES
    }
    fn live_bytes(&mut self) -> u64 {
        // A cache holds only live entries, so resident == live: amplification 1x
        // is the healthy shape (every resident byte is one the cache chose to keep).
        self.cache.len() as u64 * ENTRY_BYTES
    }
    fn structures(&mut self) -> Vec<(String, u64)> {
        vec![("entries".to_string(), self.cache.len() as u64)]
    }
    fn expected(&self) -> (SubMsGrowthClass, f64) {
        // Resident memory must never exceed the capacity's worth, no matter how
        // many distinct keys stream through. 5% slack for bookkeeping estimate.
        (
            SubMsGrowthClass::Bounded,
            (self.capacity as u64 * ENTRY_BYTES) as f64 * 1.05,
        )
    }
}

fn parse_usize(map: &BTreeMap<String, String>, key: &str, default: usize) -> usize {
    map.get(key)
        .and_then(|v| v.trim().parse().ok())
        .unwrap_or(default)
}

fn main() -> ExitCode {
    let mut raw = String::new();
    if io::stdin().read_to_string(&mut raw).is_err() {
        eprintln!("growth_main: failed to read stdin");
        return ExitCode::FAILURE;
    }
    let mut map = BTreeMap::new();
    for line in raw.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        if let Some((k, v)) = line.split_once('=') {
            map.insert(k.trim().to_string(), v.trim().to_string());
        }
    }
    let rounds = parse_usize(&map, "rounds", 50);
    let capacity = parse_usize(&map, "capacity", 1024);
    let inserts_per_round = parse_usize(&map, "inserts_per_round", 20_000);

    let mut recipe = CacheChurn {
        cache: BlockCache::with_capacity(capacity),
        capacity,
        rounds,
        inserts_per_round,
        value: vec![0u8; VALUE_BYTES],
        next: 0,
    };
    let report = grow(&mut recipe, "rust");

    if growth_to_json(&report, &mut io::stdout().lock()).is_err() {
        eprintln!("growth_main: failed to write json");
        return ExitCode::FAILURE;
    }
    ExitCode::SUCCESS
}

File: recipes/subms-block-cache/java/src/main/java/com/submillisecond/recipes/blockcache/GrowthMain.java
package com.submillisecond.recipes.blockcache;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

import com.submillisecond.perf.SubMsGrowth;
import com.submillisecond.perf.SubMsGrowthRecipe;

/**
 * Storage-growth capture for the block cache: stream far more distinct keys
 * through a fixed-capacity LRU than it can hold, and confirm the resident set
 * stays pinned at the capacity. Eviction is what bounds memory here - this
 * proves it, rather than the cache being a map that quietly grows.
 *
 * <p>Emits the stable subms growth JSON on stdout. Mirror of the Rust
 * {@code examples/growth_main.rs}, reading the same stdin key=value config, so
 * the two curves are directly comparable.
 *
 * <pre>
 * printf 'rounds=50\ncapacity=1024\ninserts_per_round=20000\n' \
 *   | java -cp target/classes:... com.submillisecond.recipes.blockcache.GrowthMain
 * </pre>
 */
public final class GrowthMain {

    private static final int VALUE_BYTES = 256;
    /** Rough per-entry heap cost: value + key (long) + slot/index bookkeeping. */
    private static final long ENTRY_BYTES = VALUE_BYTES + 8 + 32;

    private GrowthMain() {}

    private static final class CacheChurn implements SubMsGrowthRecipe {
        private final BlockCache<Long, byte[]> cache;
        private final int capacity;
        private final int rounds;
        private final int insertsPerRound;
        private final byte[] value;
        private long next;

        CacheChurn(int capacity, int rounds, int insertsPerRound) {
            this.cache = new BlockCache<>(capacity);
            this.capacity = capacity;
            this.rounds = rounds;
            this.insertsPerRound = insertsPerRound;
            this.value = new byte[VALUE_BYTES];
        }

        @Override public String name() {
            return "subms-block-cache";
        }

        @Override public String opName() {
            return "put";
        }

        @Override public int rounds() {
            return rounds;
        }

        @Override public int opsPerRound() {
            return insertsPerRound;
        }

        @Override public void op(int round, int i) {
            // A fresh distinct key every op: nothing is ever re-hit, so a naive
            // unbounded map would grow to millions - the LRU must evict to stay
            // flat.
            cache.put(next, value.clone());
            next++;
        }

        @Override public long memoryBytes() {
            return (long) cache.size() * ENTRY_BYTES;
        }

        @Override public long liveBytes() {
            // A cache holds only live entries, so resident == live: amplification
            // 1x is the healthy shape.
            return (long) cache.size() * ENTRY_BYTES;
        }

        @Override public Map<String, Long> structures() {
            return Map.of("entries", (long) cache.size());
        }

        @Override public SubMsGrowth.GrowthClass expectedClass() {
            return SubMsGrowth.GrowthClass.BOUNDED;
        }

        @Override public double expectedBound() {
            // Resident memory must never exceed the capacity's worth, no matter
            // how many distinct keys stream through. 5% slack for the bookkeeping
            // estimate.
            return (double) (capacity * ENTRY_BYTES) * 1.05;
        }
    }

    private static int parseInt(Map<String, String> cfg, String key, int fallback) {
        String v = cfg.get(key);
        if (v == null) {
            return fallback;
        }
        try {
            return Integer.parseInt(v.trim());
        } catch (NumberFormatException e) {
            return fallback;
        }
    }

    public static void main(String[] args) throws Exception {
        Map<String, String> cfg = new HashMap<>();
        try (BufferedReader in =
                new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8))) {
            String line;
            while ((line = in.readLine()) != null) {
                line = line.trim();
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }
                int eq = line.indexOf('=');
                if (eq > 0) {
                    cfg.put(line.substring(0, eq).trim(), line.substring(eq + 1).trim());
                }
            }
        }

        var recipe = new CacheChurn(
                parseInt(cfg, "capacity", 1024),
                parseInt(cfg, "rounds", 50),
                parseInt(cfg, "inserts_per_round", 20_000));

        Writer out = new OutputStreamWriter(System.out, StandardCharsets.UTF_8);
        SubMsGrowth.growthToJson(SubMsGrowth.grow(recipe, "java"), out);
        out.flush();
    }
}
